### PR TITLE
Avoid manually pinning spans in crypto interop

### PR DESCRIPTION
**Pull request by stephentoub**
My recent PRs to add span-based overloads to crypto types employed a pattern where span-based interop signatures pinned the spans manually and then passed pointers to the actual P/Invoke declaration accepting a `byte*`.  We can instead let the runtime handle the pinning by declaring the P/Invokes in terms of `ref byte`.  In addition to reducing the code involved, there's also the tiny benefit in several cases of the data being pinned for less time.

cc: @bartonjs, @ektrah

### CODE REVIEW

File: src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.Unix.cs [hunk 1]
```diff
@@ -135,11 +132,8 @@ public unsafe HmacHashProvider(IntPtr algorithmEvp, byte[] key)
                     throw new CryptographicException();
                 }
 
-                fixed (byte* keyPtr = key)
-                {
-                    _hmacCtx = Interop.Crypto.HmacCreate(keyPtr, key.Length, algorithmEvp);
-                    Interop.Crypto.CheckValidOpenSslHandle(_hmacCtx);
-                }
+                _hmacCtx = Interop.Crypto.HmacCreate(ref new Span<byte>(key).DangerousGetPinnableReference(), key.Length, algorithmEvp);
```
Comment (ektrah) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.Unix.cs]:
```csharp
_hmacCtx = Interop.Crypto.HmacCreate(ref key[0], key.Length, algorithmEvp);
```

?

File: src/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Digest.cs [hunk 2]
```diff
@@ -16,27 +16,17 @@ internal static partial class AppleCrypto
...
 
-        internal static unsafe int DigestUpdate(SafeDigestCtxHandle ctx, ReadOnlySpan<byte> pbData, int cbData)
-        {
-            fixed (byte* pbDataPtr = &pbData.DangerousGetPinnableReference())
-            {
-                return DigestUpdate(ctx, pbDataPtr, cbData);
-            }
-        }
+        internal static int DigestUpdate(SafeDigestCtxHandle ctx, ReadOnlySpan<byte> pbData, int cbData) =>
+            DigestUpdate(ctx, ref pbData.DangerousGetPinnableReference(), cbData);
```
Comment (ektrah) [inline src/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Digest.cs]: If `cbData` is always equal to `pbData.Length`, you could change this (and similar places) to

```csharp
internal static int DigestUpdate(SafeDigestCtxHandle ctx, ReadOnlySpan<byte> pbData) =>
    DigestUpdate(ctx, ref pbData.DangerousGetPinnableReference(), pbData.Length);
```

Comment (bartonjs) [inline src/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Digest.cs]: I'm pretty sure there's inconsistent style to that point; but there should be some precedent.  Early on I was exposing directly what the shim needed, then I started doing wrappers like that for legibility, convenience, and safety.

File: src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.RAND.cs [hunk 3]
```diff
@@ -9,16 +9,15 @@ internal static partial class Interop
 {
     internal static partial class Crypto
     {
-        internal static unsafe bool GetRandomBytes(byte* pbBuffer, int count)
+        internal static bool GetRandomBytes(ref byte pbBuffer, int count)
         {
-            Debug.Assert(pbBuffer != null);
```
Comment (bartonjs) [inline src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.RAND.cs]: I thought `ref byte` could be null?  So wouldn't the assert still apply?

File: src/System.Net.Http/src/System/Net/Http/ByteArrayHelpers.cs [hunk 4]
```diff
@@ -43,7 +43,7 @@ internal static bool EqualsOrdinalAsciiIgnoreCase(string left, ReadOnlySpan<byte
 
         internal static unsafe string GetStringFromByteSpan(ReadOnlySpan<byte> bytes)
         {
-            // TODO: Use new Span-based Encoding overload when available
+            // TODO #22431: Use new Span-based Encoding overload when available
```
Comment (bartonjs) [inline src/System.Net.Http/src/System/Net/Http/ByteArrayHelpers.cs]: Doesn't really belong in this PR.

File: src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.Unix.cs (hunk 1 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.Unix.cs]: Key could be the empty array, so `ref key[0]` is out of bounds.

File: src/System.Net.Http/src/System/Net/Http/ByteArrayHelpers.cs (hunk 4 above)
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/ByteArrayHelpers.cs]: I found this because I was searching for all places we were using DangerousGetPinnableReference.  While it's not directly related, didn't seem worthwhile to submit a separate PR to add 7 characters to a comment :)

File: src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.RAND.cs (hunk 3 above)
Comment (stephentoub) [inline src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.RAND.cs]: Checking `pbBuffer != null` dereferences pbBuffer and compares the value against null... which would be comparing a byte value against null, not the ref.

File: src/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Digest.cs (hunk 2 above)
Comment (stephentoub) [inline src/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Digest.cs]: I've been leaving these defined like this under the assumption that eventually we'll just be able to declare the P/Invokes with `ReadOnlySpan<byte>` instead of `ref byte` or `byte*` or `byte[]`.

File: src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.Unix.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.Unix.cs]: > Key could be the empty array, so ref key[0] is out of bounds.

Exactly.

File: src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.RAND.cs (hunk 3 above)
Comment (bartonjs) [inline src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.RAND.cs]: Yeah, but isn't there something akin to `Debug.Assert(ref pbBuffer != null)`? The assert previously was showing that we weren't shimming null, now we don't have that assertion.

Comment (ektrah) [inline src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.RAND.cs]: As far as I know it's impossible to craft a `ref byte` in C# that is a null pointer (without resorting to unsafe code).

**Comment by stephentoub** [comment]
@dotnet-bot test OSX x64 Debug Build please (https://github.com/dotnet/corefx/issues/23201)

File: src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.RAND.cs (hunk 3 above)
Comment (bartonjs) [inline src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.RAND.cs]: Interesting.  I thought the thing that sparked this was a suggestion that `ref byte` instead of `byte*` could eliminate an `unsafe` from a place passing `null`.  Maybe it was just a general suggestion and I misapplied it :smile:.

Comment (stephentoub) [inline src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.RAND.cs]: > a suggestion that ref byte instead of byte* could eliminate an unsafe from a place passing null

With unsafe you can do almost anything.  We would also need ptrs and unsafe code to write such an assert.
```C#
using System.Diagnostics;

class Program
{
    static unsafe void Main()
    {
        byte* b = null;
        Bar(ref *b);
    }

    static unsafe void Bar(ref byte b)
    {
        fixed (byte* bPtr = &b) Debug.Assert(bPtr != null);
    }
}
```

**Comment by stephentoub** [comment]
@dotnet-bot test OSX x64 Debug Build please (#23201)
